### ISSUE TITLE
Ansible installation fails due to an error with `jsmin~=2.2.2`

### ISSUE THREAD

**Issue by gregclermont** [christophetd/Adaz#44]
Hi @christophetd! Thanks for making this awesome tool!

I ran into this issue during setup:

**What is not working?**
Ansible installation fails due to an error with `jsmin~=2.2.2` (a dependency of `azure-cli==2.5.1`).

```
$ pip install -r ansible/requirements.txt
…
Collecting jsmin~=2.2.2
…
  error in jsmin setup command: use_2to3 is invalid.
…
ERROR: Could not find a version that satisfies the requirement jsmin~=2.2.2 (from azure-cli) 
```

**Cause of the error**
Quoting from https://github.com/Azure/azure-cli/issues/19468#issuecomment-914978780:

- `setuptools` deprecated `use_2to3` in `v58.0.0`
- `azure-cli` requires `jsmin`
- `jsmin` requires `use_2to3`, so it doesn't work with the latest `setuptools`
 
This new version of `setuptools` seems to be bundled in Python `3.9.8`, `3.10.1`, and all version released since Nov 2021.

**Workaround**
A quick workaround is to pin `setuptools` to a version `<58`:
```
pip install setuptools==57.5.0
```
The long-term fix is likely to update `azure-cli` to a version `>=2.28.1`, which is [when the bug was fixed](https://github.com/MicrosoftDocs/azure-docs-cli/blob/main/docs-ref-conceptual/release-notes-azure-cli.md#september-09-2021).

**What OS are you using?**
macOS 11.6

<details>
  <summary><b>Full output</b></summary>
<pre>
~/src $ git clone https://github.com/christophetd/Adaz.git
Cloning into 'Adaz'...
remote: Enumerating objects: 190, done.
remote: Counting objects: 100% (25/25), done.
remote: Compressing objects: 100% (24/24), done.
remote: Total 190 (delta 9), reused 2 (delta 0), pack-reused 165
Receiving objects: 100% (190/190), 357.57 KiB, done.
Resolving deltas: 100% (49/49), done.
~/src $ cd Adaz
~/src/Adaz (master) $ python3 -m venv ansible/venv
~/src/Adaz (master) $ source ansible/venv/bin/activate
~/src/Adaz (master) $ pip install -r ansible/requirements.txt --no-cache-dir
Collecting pywinrm==0.4.1
  Downloading pywinrm-0.4.1.tar.gz (36 kB)
  Preparing metadata (setup.py) ... done
Collecting requests==2.23.0
  Downloading requests-2.23.0-py2.py3-none-any.whl (58 kB)
Collecting msrest==0.6.13
  Downloading msrest-0.6.13-py2.py3-none-any.whl (83 kB)
Collecting msrestazure==0.6.3
  Downloading msrestazure-0.6.3-py2.py3-none-any.whl (40 kB)
Collecting azure-cli==2.5.1
  Downloading azure_cli-2.5.1-py3-none-any.whl (1.6 MB)
Collecting ansible==2.9.9
  Downloading ansible-2.9.9.tar.gz (14.2 MB)
  Preparing metadata (setup.py) ... done
Collecting xmltodict
  Downloading xmltodict-0.12.0-py2.py3-none-any.whl (9.2 kB)
Collecting requests_ntlm>=0.3.0
  Downloading requests_ntlm-1.1.0-py2.py3-none-any.whl (5.7 kB)
Collecting six
  Downloading six-1.16.0-py2.py3-none-any.whl (11 kB)
Collecting urllib3!=1.25.0,!=1.25.1,<1.26,>=1.21.1
  Downloading urllib3-1.25.11-py2.py3-none-any.whl (127 kB)
Collecting idna<3,>=2.5
  Downloading idna-2.10-py2.py3-none-any.whl (58 kB)
Collecting chardet<4,>=3.0.2
  Downloading chardet-3.0.4-py2.py3-none-any.whl (133 kB)
Collecting certifi>=2017.4.17
  Downloading certifi-2021.10.8-py2.py3-none-any.whl (149 kB)
Collecting requests-oauthlib>=0.5.0
  Downloading requests_oauthlib-1.3.0-py2.py3-none-any.whl (23 kB)
Collecting isodate>=0.6.0
  Downloading isodate-0.6.1-py2.py3-none-any.whl (41 kB)
Collecting adal<2.0.0,>=0.6.0
  Downloading adal-1.2.7-py2.py3-none-any.whl (55 kB)
Collecting azure-mgmt-eventhub~=3.0.0
  Downloading azure_mgmt_eventhub-3.0.0-py2.py3-none-any.whl (95 kB)
Collecting azure-mgmt-recoveryservicesbackup~=0.6.0
  Downloading azure_mgmt_recoveryservicesbackup-0.6.0-py2.py3-none-any.whl (194 kB)
Collecting azure-mgmt-marketplaceordering~=0.1
  Downloading azure_mgmt_marketplaceordering-0.2.1-py2.py3-none-any.whl (21 kB)
Collecting scp~=0.13.2
  Downloading scp-0.13.6-py2.py3-none-any.whl (8.2 kB)
Collecting azure-mgmt-hdinsight~=1.4.0
  Downloading azure_mgmt_hdinsight-1.4.0-py2.py3-none-any.whl (54 kB)
Collecting azure-mgmt-media>=1.1.1,~=1.1
  Downloading azure_mgmt_media-1.1.1-py2.py3-none-any.whl (341 kB)
Collecting azure-mgmt-devtestlabs~=2.2
  Downloading azure_mgmt_devtestlabs-2.2.0-py2.py3-none-any.whl (194 kB)
Collecting azure-mgmt-relay~=0.1.0
  Downloading azure_mgmt_relay-0.1.0-py2.py3-none-any.whl (36 kB)
Collecting azure-keyvault~=1.1
  Downloading azure_keyvault-1.1.0-py2.py3-none-any.whl (352 kB)
Collecting azure-mgmt-reservations==0.6.0
  Downloading azure_mgmt_reservations-0.6.0-py2.py3-none-any.whl (33 kB)
Collecting azure-batch~=9.0
  Downloading azure_batch-9.0.0-py2.py3-none-any.whl (229 kB)
Collecting azure-mgmt-security~=0.1.0
  Downloading azure_mgmt_security-0.1.0-py2.py3-none-any.whl (144 kB)
Collecting azure-cli-telemetry<2.0,>=1.0.2
  Downloading azure_cli_telemetry-1.0.6-py3-none-any.whl (10 kB)
Collecting azure-mgmt-compute~=12.0
  Downloading azure_mgmt_compute-12.1.0-py2.py3-none-any.whl (1.6 MB)
Collecting paramiko<3.0.0,>=2.0.8
  Downloading paramiko-2.9.2-py2.py3-none-any.whl (210 kB)
Collecting azure-mgmt-containerregistry~=3.0.0rc11
  Downloading azure_mgmt_containerregistry-3.0.0rc17-py2.py3-none-any.whl (660 kB)
Collecting azure-mgmt-containerservice~=9.0.1
  Downloading azure_mgmt_containerservice-9.0.1-py2.py3-none-any.whl (499 kB)
Collecting azure-mgmt-authorization~=0.52.0
  Downloading azure_mgmt_authorization-0.52.0-py2.py3-none-any.whl (112 kB)
Collecting azure-graphrbac~=0.60.0
  Downloading azure_graphrbac-0.60.0-py2.py3-none-any.whl (139 kB)
Collecting azure-mgmt-containerinstance~=1.4
  Downloading azure_mgmt_containerinstance-1.5.0-py2.py3-none-any.whl (96 kB)
Collecting azure-mgmt-botservice~=0.2.0
  Downloading azure_mgmt_botservice-0.2.0-py2.py3-none-any.whl (109 kB)
Collecting azure-mgmt-sql~=0.18.0
  Downloading azure_mgmt_sql-0.18.0-py2.py3-none-any.whl (375 kB)
Collecting azure-mgmt-managementgroups~=0.1
  Downloading azure_mgmt_managementgroups-0.2.0-py2.py3-none-any.whl (59 kB)
Collecting mock~=4.0
  Downloading mock-4.0.3-py3-none-any.whl (28 kB)
Collecting azure-mgmt-web~=0.44.0
  Downloading azure_mgmt_web-0.44.0-py2.py3-none-any.whl (1.0 MB)
Collecting urllib3[secure]~=1.18
  Downloading urllib3-1.26.8-py2.py3-none-any.whl (138 kB)
Collecting azure-mgmt-servicefabric~=0.4.0
  Downloading azure_mgmt_servicefabric-0.4.0-py2.py3-none-any.whl (62 kB)
Collecting azure-mgmt-trafficmanager~=0.51.0
  Downloading azure_mgmt_trafficmanager-0.51.0-py2.py3-none-any.whl (58 kB)
Collecting azure-mgmt-datalake-store~=0.5.0
  Downloading azure_mgmt_datalake_store-0.5.0-py2.py3-none-any.whl (88 kB)
Collecting azure-mgmt-iotcentral~=3.0.0
  Downloading azure_mgmt_iotcentral-3.0.0-py2.py3-none-any.whl (18 kB)
Collecting azure-mgmt-cdn==4.1.0rc1
  Downloading azure_mgmt_cdn-4.1.0rc1-py2.py3-none-any.whl (73 kB)
Collecting azure-mgmt-signalr~=0.3.0
  Downloading azure_mgmt_signalr-0.3.0-py2.py3-none-any.whl (53 kB)
Collecting jsondiff==1.2.0
  Downloading jsondiff-1.2.0.tar.gz (8.0 kB)
  Preparing metadata (setup.py) ... done
Collecting azure-mgmt-consumption~=2.0
  Downloading azure_mgmt_consumption-2.0.0-py2.py3-none-any.whl (46 kB)
Collecting azure-mgmt-deploymentmanager~=0.2.0
  Downloading azure_mgmt_deploymentmanager-0.2.0-py2.py3-none-any.whl (41 kB)
Collecting azure-mgmt-redis~=7.0.0rc1
  Downloading azure_mgmt_redis-7.0.0-py2.py3-none-any.whl (44 kB)
Collecting azure-mgmt-managedservices~=1.0
  Downloading azure_mgmt_managedservices-1.0.0-py2.py3-none-any.whl (38 kB)
Collecting azure-mgmt-policyinsights~=0.4.0
  Downloading azure_mgmt_policyinsights-0.4.0-py2.py3-none-any.whl (41 kB)
Collecting sshtunnel~=0.1.4
  Downloading sshtunnel-0.1.5-py2.py3-none-any.whl (23 kB)
Collecting azure-mgmt-appconfiguration~=0.4.0
  Downloading azure_mgmt_appconfiguration-0.4.0-py2.py3-none-any.whl (28 kB)
Collecting azure-mgmt-keyvault~=2.2.0
  Downloading azure_mgmt_keyvault-2.2.0-py2.py3-none-any.whl (89 kB)
Collecting vsts-cd-manager>=1.0.2,~=1.0.0
  Downloading vsts-cd-manager-1.0.2.tar.gz (13 kB)
  Preparing metadata (setup.py) ... done
Collecting azure-mgmt-cognitiveservices~=5.0.0
  Downloading azure_mgmt_cognitiveservices-5.0.0-py2.py3-none-any.whl (31 kB)
Collecting azure-mgmt-imagebuilder~=0.2.1
  Downloading azure_mgmt_imagebuilder-0.2.1-py2.py3-none-any.whl (66 kB)
Collecting azure-cosmos>=3.0.2,~=3.0
  Downloading azure_cosmos-3.2.0-py2.py3-none-any.whl (106 kB)
Collecting pyOpenSSL>=17.1.0
  Downloading pyOpenSSL-21.0.0-py2.py3-none-any.whl (55 kB)
Collecting azure-mgmt-storage~=9.0.0
  Downloading azure_mgmt_storage-9.0.0-py2.py3-none-any.whl (525 kB)
Collecting azure-mgmt-batchai~=2.0
  Downloading azure_mgmt_batchai-2.0.0-py2.py3-none-any.whl (174 kB)
Collecting azure-mgmt-iothub~=0.11.0
  Downloading azure_mgmt_iothub-0.11.0-py2.py3-none-any.whl (401 kB)
Collecting azure-mgmt-dns~=2.1
  Downloading azure_mgmt_dns-2.1.0-py2.py3-none-any.whl (134 kB)
Collecting azure-mgmt-msi~=0.2
  Downloading azure_mgmt_msi-0.2.0-py2.py3-none-any.whl (17 kB)
Collecting azure-mgmt-billing~=0.2
  Downloading azure_mgmt_billing-0.2.0-py2.py3-none-any.whl (25 kB)
Collecting javaproperties==0.5.1
  Downloading javaproperties-0.5.1-py2.py3-none-any.whl (19 kB)
Collecting azure-multiapi-storage~=0.3.1
  Downloading azure_multiapi_storage-0.3.7-py2.py3-none-any.whl (2.4 MB)
Collecting azure-mgmt-batch~=7.0
  Downloading azure_mgmt_batch-7.0.0-py2.py3-none-any.whl (81 kB)
Collecting azure-mgmt-recoveryservices~=0.4.0
  Downloading azure_mgmt_recoveryservices-0.4.0-py2.py3-none-any.whl (77 kB)
Collecting azure-cli-command-modules-nspkg~=2.0
  Downloading azure_cli_command_modules_nspkg-2.0.3-py3-none-any.whl (1.9 kB)
Collecting azure-mgmt-servicebus~=0.6.0
  Downloading azure_mgmt_servicebus-0.6.0-py2.py3-none-any.whl (120 kB)
Collecting azure-loganalytics~=0.1.0
  Downloading azure_loganalytics-0.1.1-py2.py3-none-any.whl (16 kB)
Collecting cryptography<3.0.0,>=2.3.1
  Downloading cryptography-2.9.2-cp35-abi3-macosx_10_9_x86_64.whl (1.8 MB)
Collecting azure-mgmt-kusto~=0.3.0
  Downloading azure_mgmt_kusto-0.3.0-py2.py3-none-any.whl (73 kB)
Collecting azure-cli-nspkg>=2.0.0
  Downloading azure_cli_nspkg-3.0.4-py3-none-any.whl (1.8 kB)
Collecting azure-mgmt-sqlvirtualmachine~=0.5.0
  Downloading azure_mgmt_sqlvirtualmachine-0.5.0-py2.py3-none-any.whl (34 kB)
Collecting azure-storage-blob<2.0.0,>=1.3.1
  Downloading azure_storage_blob-1.5.0-py2.py3-none-any.whl (75 kB)
Collecting azure-mgmt-loganalytics~=0.5.0
  Downloading azure_mgmt_loganalytics-0.5.0-py2.py3-none-any.whl (71 kB)
Collecting azure-mgmt-datalake-analytics~=0.2.1
  Downloading azure_mgmt_datalake_analytics-0.2.1-py2.py3-none-any.whl (146 kB)
Collecting azure-functions-devops-build~=0.0.22
  Downloading azure_functions_devops_build-0.0.22-py3-none-any.whl (47 kB)
Collecting azure-mgmt-resource==9.0.0
  Downloading azure_mgmt_resource-9.0.0-py2.py3-none-any.whl (807 kB)
Collecting antlr4-python3-runtime~=4.7.2
  Downloading antlr4-python3-runtime-4.7.2.tar.gz (112 kB)
  Preparing metadata (setup.py) ... done
Collecting websocket-client~=0.56.0
  Downloading websocket_client-0.56.0-py2.py3-none-any.whl (200 kB)
Collecting azure-mgmt-redhatopenshift==0.1.0
  Downloading azure_mgmt_redhatopenshift-0.1.0-py2.py3-none-any.whl (23 kB)
Collecting azure-mgmt-eventgrid~=2.2
  Downloading azure_mgmt_eventgrid-2.2.0-py2.py3-none-any.whl (109 kB)
Collecting azure-mgmt-iothubprovisioningservices~=0.2.0
  Downloading azure_mgmt_iothubprovisioningservices-0.2.0-py2.py3-none-any.whl (60 kB)
Collecting azure-mgmt-rdbms~=2.2.0
  Downloading azure_mgmt_rdbms-2.2.0-py2.py3-none-any.whl (231 kB)
Collecting azure-cli-core==2.5.1.*
  Downloading azure_cli_core-2.5.1-py3-none-any.whl (140 kB)
Collecting azure-mgmt-privatedns~=0.1.0
  Downloading azure_mgmt_privatedns-0.1.0-py2.py3-none-any.whl (44 kB)
Collecting azure-mgmt-netapp~=0.8.0
  Downloading azure_mgmt_netapp-0.8.0-py2.py3-none-any.whl (40 kB)
Collecting azure-mgmt-apimanagement~=0.1.0
  Downloading azure_mgmt_apimanagement-0.1.0-py2.py3-none-any.whl (542 kB)
Collecting azure-mgmt-network~=10.1.0
  Downloading azure_mgmt_network-10.1.0-py2.py3-none-any.whl (8.1 MB)
Collecting pytz==2019.1
  Downloading pytz-2019.1-py2.py3-none-any.whl (510 kB)
Collecting jsmin~=2.2.2
  Downloading jsmin-2.2.2.tar.gz (12 kB)
  Preparing metadata (setup.py) ... error
  ERROR: Command errored out with exit status 1:
   command: /Users/user/src/Adaz/ansible/venv/bin/python3.9 -c 'import io, os, sys, setuptools, tokenize; sys.argv[0] = '"'"'/private/var/folders/yj/m9_s7pl516jgd8wvwb0snqk40000gn/T/pip-install-e7uhszi_/jsmin_18eb338660e6417dba5daa1ee53be54a/setup.py'"'"'; __file__='"'"'/private/var/folders/yj/m9_s7pl516jgd8wvwb0snqk40000gn/T/pip-install-e7uhszi_/jsmin_18eb338660e6417dba5daa1ee53be54a/setup.py'"'"';f = getattr(tokenize, '"'"'open'"'"', open)(__file__) if os.path.exists(__file__) else io.StringIO('"'"'from setuptools import setup; setup()'"'"');code = f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' egg_info --egg-base /private/var/folders/yj/m9_s7pl516jgd8wvwb0snqk40000gn/T/pip-pip-egg-info-mfhzutpd
       cwd: /private/var/folders/yj/m9_s7pl516jgd8wvwb0snqk40000gn/T/pip-install-e7uhszi_/jsmin_18eb338660e6417dba5daa1ee53be54a/
  Complete output (1 lines):
  error in jsmin setup command: use_2to3 is invalid.
  ----------------------------------------
WARNING: Discarding https://files.pythonhosted.org/packages/17/73/615d1267a82ed26cd7c124108c3c61169d8e40c36d393883eaee3a561852/jsmin-2.2.2.tar.gz#sha256=b6df99b2cd1c75d9d342e4335b535789b8da9107ec748212706ef7bbe5c2553b (from https://pypi.org/simple/jsmin/). Command errored out with exit status 1: python setup.py egg_info Check the logs for full command output.
INFO: pip is looking at multiple versions of msrestazure to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of msrest to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of <Python from Requires-Python> to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of requests to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of pywinrm to determine which version is compatible with other requirements. This could take a while.
ERROR: Could not find a version that satisfies the requirement jsmin~=2.2.2 (from azure-cli) (from versions: 2.0, 2.0.1, 2.0.2, 2.0.2.post1, 2.0.3, 2.0.5, 2.0.6, 2.0.7, 2.0.8, 2.0.9, 2.0.10, 2.0.11, 2.1.0, 2.1.1, 2.1.2, 2.1.3, 2.1.4, 2.1.5, 2.1.6, 2.2.0, 2.2.1, 2.2.2, 3.0.0)
ERROR: No matching distribution found for jsmin~=2.2.2
~/src/Adaz (master) $ pip list
Package    Version
---------- -------
pip        21.3.1
setuptools 59.0.1
~/src/Adaz (master) $ python3 --version
Python 3.9.9
</pre>
</details>

**Comment by gregclermont** [comment]
Is `azure-cli` used for anything else than the authentication? If not, it might be easy to update.

**Comment by christophetd** [comment]
Hello,

Thanks for reporting! Can you give it a try with this patch? I think you're right and we don't need to include the Azure CLI as part of this project's dependencies:

```diff
diff --git a/ansible/requirements.txt b/ansible/requirements.txt
index 6bf7b8b..a883102 100644
--- a/ansible/requirements.txt
+++ b/ansible/requirements.txt
@@ -1,6 +1,5 @@
 pywinrm==0.4.1
 requests==2.23.0
 msrest==0.6.13
-msrestazure==0.6.3
-azure-cli==2.5.1
+msrestazure==0.6.4
 ansible==2.9.9

```

**Comment by gregclermont** [comment]
Hi! Sorry for the delay.
I have tested with the patch, and I have found another issue:

The project uses the provider `azurerm v2.12.0`.
https://github.com/christophetd/Adaz/blob/06d0264130fa4156c43432a899781efa1d25ba6d/terraform/provider.tf#L1-L2

A breaking change was introduced last November in `Azure CLI 2.30.0` (some details in [this issue](https://github.com/hashicorp/terraform-provider-azurerm/issues/14560)), which means that recent versions of the Azure CLI aren't compatible with older versions of `azurerm` (I didn' t manage to find in which version the fix was made).

To make it work, I've had to install the last compatible release via `pip`, `azure-cli==2.29.2`.
